### PR TITLE
Add iCloud Drive support

### PR DESCRIPTION
**Pull request by saintamh**
This PR adds a new service, `api.drive`, which provides access to  files stored in iCloud drive.

The interface fpr the new service class is a copy of the Ubiquity module. I thought it makes sense to keep things coherent. I'm happy to abstract away the commonality into a `Filesystem` class that both Ubiquity and Drive would inherit from, if that's judged preferable.

### CODE REVIEW

**Comment by torarnv** [comment]
Nice. Why would we need the Ubiquity module though? Doesn't this replace it 1:1?

**Comment by saintamh** [comment]
@torarnv I'm not sure how much this replaces the Ubiquity service, I'm not really familiar with how Ubiquity works. What I can see is that, logging in with my personal account, I see an empty directory in Ubiquity (even though I have e.g. notes from the iOS "Notes" app saved to the cloud). But if I save files to my iCloud Drive, the Drive service in this PR lets me access them.

```
>>> api.files.dir()
['.do-not-delete']
>>> api.drive.dir()
['files-i-uploaded-manually']
```

I've only tested all this with my own account, and in my case, the services don't seem to overlap. But it may be that if I did have stuff in `api.files`, it would also show up under `api.drive`, I don't know.

File: pyicloud/services/drive.py [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+                    ))
+                return match.group(1)
+        raise Exception("Token cookie not found")
+
+    def get_file(self, id, **kwargs):
+        meta = self.session.get(
+            self._service_root + '/ws/com.apple.CloudDocs/download/by_id',
+            params={
+                'document_id': id,
+                'token': self.get_token_from_cookie(),
```
Comment (torarnv) [inline pyicloud/services/drive.py]: The token is set on auth right? So you should be able to store it in __init__?

File: pyicloud/base.py [hunk 2]
```diff
@@ -84,20 +85,21 @@ def request(self, *args, **kwargs):
...
         logger.debug(json)
 
-        reason = json.get('errorMessage')
-        reason = reason or json.get('reason')
-        reason = reason or json.get('errorReason')
-        if not reason and isinstance(json.get('error'), six.string_types):
-            reason = json.get('error')
-        if not reason and json.get('error'):
-            reason = "Unknown reason"
+        if isinstance(json, dict):
```
Comment (torarnv) [inline pyicloud/base.py]: This change seems to be unrelated, why is it needed?

Comment (saintamh) [reply pyicloud/base.py]: For the Drive requests, `json` is not a dict, it's a list, which evidently is not something that had been encountered before. Either the API works differently, or I'm not making the right call?

File: pyicloud/services/drive.py (hunk 1 above)
Comment (saintamh) [reply pyicloud/services/drive.py]: That's a good point I'll make that change.

Comment (saintamh) [reply pyicloud/services/drive.py]: Done

**Comment by WANG-lp** [comment]
Any idea? It seems reasonable to implement the iCloud-Drive interface

**Comment by mohd-akram** [comment]
Any update on this?

**Comment by Quentame** [comment]
I will test if working during this or next week and will see if I can merge or should update some code before.

**Comment by Quentame** [comment]
Hi @saintamh !

Can you still work on this or I should take the lead ?

Thanks.

**Comment by saintamh** [comment]
Go for it @Quentame. I won't have time for this at all these days unfortunately.

**Comment by Quentame** [comment]
Merged, thanks a lot @saintamh !

I've added you as co-author :wink:
